### ISSUE TITLE
Reproducibility

### ISSUE THREAD

**Issue by AmirMardan** [AmirMardan/pinn_fwi#4]
The result is not reproducible when we use multiscale fwi, seems there is a bug in `lpass`
https://github.com/AmirMardan/pinn_fwi/blob/main/torchfwi.py#L31-L33

**Comment by AmirMardan** [comment]
This is because of Adam and the nondeterminism in pytorch

https://discuss.pytorch.org/t/pytorch-reproducibility/135980

For the optimization, we optimize the network for every batch and every frequency,
```Python
def train(...):
    loss_data_minibatch = []
    loss_prior_minibatch = []
    for batch in range(mini_batches):
        loss_freqs = []
        loss_priors = []
        for freq in freqs:
            optim_autoencoder.zero_grad()
            
            src_loc_batch = src_loc[batch::mini_batches]
            rec_loc_batch = rec_loc[batch::mini_batches]
            src_batch = src[batch::mini_batches]
                    
            physics = Physics(inpa['dh'], inpa['dt'], src=src_batch,
                            src_loc=src_loc_batch, rec_loc=rec_loc_batch
                            )

            loss_data, loss_prior, m = train_engine(autoencoder, physics,
                                   criteria, optim_autoencoder,
                                   d_obs, freq,
                                   batch, mini_batches,
                                   lam_prior=lam_prior,
                                   well_locations=well_locations,
                                   well_data=well_data,
                                   test=test
                                   )
            loss_freqs.append(loss_data)
            loss_priors.append(loss_prior)
        
        loss_data_minibatch.append(np.mean(loss_freqs))
        loss_prior_minibatch.append(np.mean(loss_priors))
    return np.mean(loss_data_minibatch), np.mean(loss_prior_minibatch), m, autoencoder

```

In the following figures, we can see that the loss for the first batch and first frequency are equal in two different runs.

<img width="557" alt="run1" src="https://github.com/AmirMardan/pinn_fwi/assets/46511946/65cc340d-2019-4695-8668-949d593553e5">
<img width="542" alt="run2" src="https://github.com/AmirMardan/pinn_fwi/assets/46511946/d9b3f075-b101-4ad1-90d5-cf4d985f2cf6">

As is shown, the first loss value is exactly reproduced however the loss changes for the next frequencies and batches.
